gh auth login

vilmibm opened the pull request:
This PR adds `gh auth login`.

## interactive mode - paste token

![Peek 2020-07-31 17-25](https://user-images.githubusercontent.com/98482/89082242-03ef9280-d353-11ea-94f3-76bffe4322fd.gif)

## interactive mode - browser flow

![Peek 2020-07-31 17-24](https://user-images.githubusercontent.com/98482/89082236-018d3880-d353-11ea-8727-43228a74f17f.gif)



## non-interactive mode

![authloginsstdin](https://user-images.githubusercontent.com/98482/88857447-e936e580-d1bb-11ea-93b6-ed9801c2517c.gif)

Part of #1413 


# Open questions:

- I'm not sure what support for the device auth flow looks like and was hoping follow up with that later; is that reasonable?
- We're not set up to test the browser auth flow. Should I hack in support for that or continue to rely on manual testing?
- For the survey-based tests, I didn't test the printing of informational messages to STDERR. I figured just ensuring that the correct config was written given a set of prompt stubs was good enough. Is it?

vilmibm committed:
gh auth login

mislav commented on internal/config/testing.go:
```diff
@@ -23,9 +24,15 @@ func StubWriteConfig(wc io.Writer, wh io.Writer) func() {
 		switch path.Base(fn) {
 		case "config.yml":
+			if b, ok := wc.(*bytes.Buffer); ok {
+				b.Reset()
```
I'm starting the review with this file and I'm not yet sure about the context for this change, but it looks like a smell to me. Why would WriteConfigFile need to rewind the same input buffer? Should we work around that in at the call sites?

vilmibm replied:
If any command would write to a config file twice in the course of its execution the buffer would have both writes; login .Write()s twice (to be able to test a new token) and I noticed the double-written buffer. This seemed like a reasonable thing to do in tests but I'm also I'd be interested in not having to do the two writes in `login`, making this a moot point.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+			isTTY := opts.IO.IsStdinTTY()
+
+			if !isTTY {
+				if !cmd.Flags().Changed("with-token") {
```
Nit: this feels like it could be collapsed into one condition

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.Hostname, "hostname", "", "The hostname of the GitHub instance to authenticate with")
```
I would be interested in supporting the shorthand `-h` here so we can give users short-looking instructions on how to authenticate with a particular host. Or, we could name it `--host`!

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	cmd.Flags().StringVar(&opts.Hostname, "hostname", "", "The hostname of the GitHub instance to authenticate with")
+	cmd.Flags().Bool("with-token", false, "If specified, token is read from STDIN")
```
Language nit: I would like to generally avoid starting boolean flag descriptions with “If specified”, since that is implied. How about: “Read the token from standard input”

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+		if err != nil {
+			return err
+		}
+		err = cfg.Write()
```
This writes the token to config file before the token is verified. Could we first verify it and then call `Write` on success? This would depend on the API client reading the token from the Config instance in-memory, which would of course only work if the API client has access to the same Config instance as we have here 🤔

vilmibm replied:
The code as it stands doesn't support this :( but today I can try and get this to work. This is definitely a hack-around.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+
+	if opts.Token != "" {
+		if opts.Hostname == "" {
+			panic("empty hostname would leak oauth_token")
```
This is nice precaution! Could we return an error in this case?

vilmibm replied:
we could; but this felt more like something that could come up in development as opposed to a user's running code (and the panic is to prevent it from getting out of development).

I guess returning an error would have the same effect and there's no reason not to.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+			return fmt.Errorf("could not prompt: %w", err)
+		}
+
+		if hostname == "" {
+			panic("empty hostname would leak oauth_token")
+		}
+		err = cfg.Set(hostname, "oauth_token", token)
+		if err != nil {
+			return err
+		}
+		err = cfg.Write()
+		if err != nil {
+			return err
+		}
+
+		httpClient, err := opts.HttpClient()
+		if err != nil {
+			return err
+		}
+
+		apiClient := api.NewClientFromHTTP(httpClient)
+
+		hasScopes, _, err := apiClient.HasScopes(expectedScopes...)
+		if err != nil {
+			return fmt.Errorf("could not verify token: %w", err)
+		}
+
+		if !hasScopes {
```
It feels like this whole block of code is identical or near-identical as the above one from non-interactive mode. Since they are just input modes, could both modes lead towards the same code path?

vilmibm replied:
most likely! I meant to circle back to this repetition.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+		return err
+	}
+
+	apiClient := api.NewClientFromHTTP(httpClient)
```
This is the 3rd time this function initializes both `httpClient` and `apiClient`. I believe it would be better to do it only once at the top, but with the current state of trunk I'm guessing we need to keep refreshing the client whenever we change credentials?

If you base this work on https://github.com/cli/cli/pull/1415, however, then an API client doesn't store a cached token, but always consults its Config instance for one. So maybe `auth` could either take advantage of that, or maintain its own authentication strategy as an alternative!

vilmibm replied:
> but with the current state of trunk I'm guessing we need to keep refreshing the client whenever we change credentials?

yup.

I can rebase on 1415 and try to clean that up.

mislav commented on pkg/cmd/auth/login/login_test.go:
```diff
@@ -0,0 +1,328 @@
+		{
+			name: "hostname set",
+			opts: &LoginOptions{
+				Hostname: "rebecca.chambers",
```
☣️

mislav commented on pkg/prompt/stubber.go:
```diff
@@ -8,15 +8,40 @@ import (
-func InitAskStubber() (*askStubber, func()) {
+// TODO ADAPT THIS ALL TO SUPPORT THE NEW SURVEYASKONE
+
+func InitAskStubber() (*AskStubber, func()) {
```
One thing we could explore with isolated commands is moving away from this kind of per-package stubbing mechanism and passing a survey implementation as an object via a Factory. Then the test could pass in a mock survey implementation that simulates user's choices. Something to consider if the current approach becomes too hairy!

vilmibm replied:
It doesn't feel like too much yet! this TODO is stale. It's not the most elegant approach but I think it's fine for now.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+			return fmt.Errorf("failed to authenticate via web browser: %w", err)
+		}
+	} else {
+		fmt.Fprintln(opts.IO.ErrOut)
+		fmt.Fprintln(opts.IO.ErrOut, heredoc.Doc(`
+				Tip: you can generate a Personal Access Token here https://github.com/settings/tokens
+				The minimum required scopes are 'repo' and 'read:org'.`))
```
Nit, but I'd be fine with printing information like this to stdout since it's part of the essential output of the `login` command. I do not expect that people will be redirecting stdout of `login` to a script or anyting

vilmibm replied:
I'm torn; on the one hand, printing to STDOUT is absolutely fine here (and survey is using STDOUT). On the other, we print this kind of stuff to STDERR in other commands and I'd like to be consistent.

mislav replied:
> On the other, we print this kind of stuff to STDERR in other commands and I'd like to be consistent.

True. But the way I see it, the only reason we separate out "special" notices from regular output is because we expect that the user might be piping regular output to a file or a script. If the user is not doing that (for this command, they are not likely to), there is no incentive for us to print notices on a separate stream.

jsejcksn replied:
Some software writes to the tty in raw mode for these types of informational notices, so the text is neither in `stdout` nor `stderr`.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,284 @@
+
+		apiClient := api.NewClientFromHTTP(httpClient)
+
+		hasScopes, _, err := apiClient.HasScopes(expectedScopes...)
```
Something that just occurred to me: does HasScopes know that `admin:org` satisfies `read:org`? It's something that I had to teach to CheckScopes https://github.com/cli/cli/pull/1359 but I don't think it applied to HasScopes since the latter was only meant for gist back then

vilmibm replied:
ah, I don't think it would.

vilmibm replied:
oh, uh, `HasScopes` also has `api.github.com` hardcoded :joy: so when I used it to validate tokens for GHE hosts it's just a lie

mislav reviewed (changes requested):
It's great that we're finally going to have this command 🎉

mislav commented:
> * I'm not sure what support for the device auth flow looks like and was hoping follow up with that later; is that reasonable?

Yes, I think device flow should absolutely be a follow-up. The device flow can be feature-detected and fits under the "Login with a web browser” flow as far as the user is concerned, so its implementation will not require almost any UI tweaks.

> * We're not set up to test the browser auth flow. Should I hack in support for that or continue to rely on manual testing?

Uh, I declared bankruptcy on testing the browser auth flow since I did not have any ideas how to stub it back then. If you do, that would be awesome! Especially because after this, I would really like us to opensource the `auth` package so hub can use it as well.

> * For the survey-based tests, I didn't test the printing of informational messages to STDERR. I figured just ensuring that the correct config was written given a set of prompt stubs was good enough. Is it?

I think that's good enough, especially since we might be tweaking the flow still. Credentials being correctly saved is most of what matters

vilmibm commented:
note to self: do some bolding / green check / formatting on browser flow prompts

vilmibm commented:
@mislav 

I've...

- rebased on `ghe-api` branch
- properly validate scopes now (both in terms of handling admin:read and using the correct host)
- made the `login` command create its own http client from an in-memory config
- streamlined config interactions (only one write per command invocation, now)
- added check for already having a valid authentication for the given host

mislav commented on api/client.go:
```diff
@@ -218,26 +218,36 @@ func (c Client) HasScopes(wantedScopes ...string) (bool, string, error) {
+	search := map[string]bool{
+		"repo":       false,
+		"read:org":   false,
+		"admin:read": false,
```
This should be `admin:org` instead of `admin:read`

mislav commented on api/client.go:
```diff
@@ -218,26 +218,36 @@ func (c Client) HasScopes(wantedScopes ...string) (bool, string, error) {
-		return true, appID, nil
+	errorMsgs := []string{}
+	if !search["repo"] {
+		errorMsgs = append(errorMsgs, "missing required scope repo")
```
Should we maybe quote "repo" to draw attention to it and separate it out from prose?

mislav commented on api/client.go:
```diff
@@ -218,26 +218,36 @@ func (c Client) HasScopes(wantedScopes ...string) (bool, string, error) {
 
-	return false, appID, nil
+	if !search["read:org"] && !search["admin:read"] {
+		errorMsgs = append(errorMsgs, "missing a read scope (need one of read:org or admin:read)")
```
Although we are technically satisfied by `admin:org` scope, I think we should avoid ever hinting that gh _needs_ it in the error message. So I think it's best just to say `missing required scope "read:org"`

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,280 @@
+
+			wt, _ := cmd.Flags().GetBool("with-token")
+			if wt {
+				if isTTY {
+					return &cmdutil.FlagError{Err: errors.New("expected token on STDIN")}
```
If someone specified `--with-token`, then they've explicitly opted into stdin mode. I don't think we need to try to be helpful here and check the source of stdin: it could be a file, or pipe, or terminal; it should be all equally fine. I don't think we should disallow reading the token from terminal stdin, even if it's going to be echoed in terminal output.

I think so far, generally we've only errored out if the user hasn't passed any arguments/flags specifically and stdin was a tty, so their intent was ambiguous and in that case, we want to be more helpful and error out instead of blocking by taking input from tty stdin.

vilmibm replied:
good point, I forgot about just letting people paste via the tty

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,280 @@
+				if isTTY {
+					return &cmdutil.FlagError{Err: errors.New("expected token on STDIN")}
+				}
+				defer opts.IO.In.Close()
+				token, err := ioutil.ReadAll(opts.IO.In)
```
I'm not sure, but I think the logic to actually read the token from stdin might belong in `loginRun` instead of here. Typically, we used `RunE` so far to just to argument/flags processing, but avoided executing any command logic. I'm on the fence about this, so I could go either way. What do you think?

vilmibm replied:
I was very unsure. I think this falls under the category of argument parsing if you consider the value on STDIN to be an argument. I felt that the "inputs" of this command were a hostname and a token and that `loginRun` should only care about those two things and be less concerned with whence they came. I decided to see how uncomfortable the testing felt and that turned out to feel fine so I'm content to leave it like this.

mislav replied:
That reasoning makes sense!

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,280 @@
+			return err
+		}
+
+		err = validateHostCfg(opts.Hostname, cfg)
```
Love how elegant this pre-check feels before `cfg.Write()` 👍

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,280 @@
+
+	isTTY := opts.IO.IsStdoutTTY() && opts.IO.IsStdinTTY()
+
+	if !isTTY {
+		return errors.New("token must be passed via STDIN and --with-token when unattached to TTY")
```
This check and the error case feels like it could belong in flag processing in `RunE`.

Also, my brain is having difficulty parsing the error message. How about something like:
```
No terminal detected; please use `--with-token` and supply a token via standard input
```

vilmibm replied:
This checking was in `RunE`; I just accidentally duplicated it. I took out this check and reworded the other.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,280 @@
+		return errors.New("token must be passed via STDIN and --with-token when unattached to TTY")
+	}
+
+	// TODO consider explicitly telling survey what io to use since it's implicit right now
```
Great thinking! 💯

mislav commented on pkg/httpmock/stub.go:
```diff
@@ -56,6 +56,26 @@ func GraphQL(q string) Matcher {
+		if !strings.EqualFold(req.Method, "POST") {
+			return false
+		}
+		if req.URL.Path != "/api/graphql" {
```
I can tell why you needed to add an extra matcher, but I'm not sure if it's worth a new method just to match an alternate API endpoint. Could you instead amend the `GraphQL` matcher to match on either `/graphql` or `/api/graphql` indiscriminately? Most assertions should not be concerned with whether the request was "Enterprise" or not; and our API routing is already tested elsewhere.

vilmibm replied:
that makes perfect sense, i'll do that.

mislav reviewed (changes requested):
Everything looks great; thank you!

The only potential blocker from my standpoint is GITHUB_TOKEN handling. As it stands, this branch only respects configuration from `~/.config/gh/hosts.yml` file, but if a user supplied an invalid or revoked GITHUB_TOKEN via their environment, any attempt of a `gh` command will fail with a HTTP 401, but `gh auth login` will show “You're already logged into github.com as mislav. Do you want to re-authenticate?” And even if I went through re-authentication process, a new token will be written to my config file, but the old (broken) GITHUB_TOKEN from environment still takes precedence and sabotages my gh operations.

danshearer commented:
This is an excellent improvement. Thanks!

Also, the wee screencasts let me see the entire scope of what you are trying to do in seconds. Great.

mislav commented on pkg/cmd/auth/login/client.go:
```diff
@@ -0,0 +1,48 @@
+	return nil
+}
+
+var clientFromCfg = func(hostname string, cfg config.Config) (*api.Client, error) {
```
Is there a reason why this and `validateHostCfg` are overridable functions (i.e. assigned to a `var`)?

vilmibm replied:
clientFromCfg is overriden in tests; validateHostCfg is not and that's just vestigial, I'll fix that.

mislav commented on api/client.go:
```diff
@@ -195,19 +196,18 @@ func (err HTTPError) Error() string {
 
-// Returns whether or not scopes are present, appID, and error
-func (c Client) HasScopes(wantedScopes ...string) (bool, string, error) {
-	url := "https://api.github.com/user"
```
It's great that this no longer depends on the `/user` endpoint!

mislav commented on api/client.go:
```diff
@@ -218,26 +218,36 @@ func (c Client) HasScopes(wantedScopes ...string) (bool, string, error) {
 	}
 
-	return false, appID, nil
+	if !search["read:org"] && !search["admin:org"] {
```
Love the simplicity of this new implementation for checking scopes 🏅

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,281 @@
+
+				opts.Token = strings.TrimSpace(string(token))
+			} else if ghToken != "" {
+				opts.Token = ghToken
```
I think that making GITHUB_TOKEN here indistinguishable from the one passed via `--with-token` is not the behavior that I would expect. That could result in GITHUB_TOKEN being stored to `hosts.yml`, which I don't think should ever happen since GITHUB_TOKEN is environment-only.

It's useful to think of GITHUB_TOKEN as credentials stored in a virtual layer "above" (and entirely separate of) `hosts.yml`. So I think that `gh auth login/status` should be aware of GITHUB_TOKEN being set and help you determine whether the credentials are valid, but they can't help you interactively refresh the credentials since only the user can change an environment variable.

vilmibm replied:
> That could result in GITHUB_TOKEN being stored to hosts.yml, which I don't think should ever happen since GITHUB_TOKEN is environment-only.

I was going quickly and didn't even think about this :woman_facepalming: I will at least push a commit to not save the config if `GITHUB_TOKEN` is being validated.

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -0,0 +1,281 @@
+			return errors.New("empty hostname would leak oauth_token")
+		}
+
+		err := cfg.Set(opts.Hostname, "oauth_token", opts.Token)
```
This only stores `oauth_token` key for a host and not the `user` key, but maybe that's fine since we don't really need the `user` key for anything other than reassurance for people manually inspecting their config file.

vilmibm replied:
oh, that's a good point though; I might as well set that.

mislav reviewed (approved):
Just a minor point about GITHUB_TOKEN handling, but would be also fine if we addressed various authentication issues and error messaging around GITHUB_TOKEN in a follow-up. 🌟

tierninho commented:
@vilmibm Curious if we need a URL validation on this as it accepts jibberish and even a blank `space`

```
? What account do you want to log into? GitHub Enterprise
? GHE hostname: ghsdfsdf
- Logging into ghsdfsdf
? How would you like to authenticate? Login with a web browser
- Press Enter to open ghsdfsdf in your browser...
```

The app hangs as the browser just goes to a dead page as expected.

Happy to open an issue if needed.
